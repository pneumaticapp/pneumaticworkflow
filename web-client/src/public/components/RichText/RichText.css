/* stylelint-disable selector-max-compound-selectors */
.container * {
  margin: 0;
  max-width: 100%;
  font-size: inherit;
  line-height: inherit;
  white-space: pre-wrap;
}

.container {
  overflow-wrap: anywhere;

  hr {
    margin: 16px 0;
    max-width: 370px;
  }

  a {
    color: var(--pneumatic-color-link-dark);
    transition: color 0.1s ease-in-out;

    &:hover {
      color: var(--pneumatic-color-link-hover);
    }
  }

  ul,
  ol {
    padding-left: 32px;
    list-style-position: inside;

    li::marker {
      font-weight: bold;
    }

    p:first-of-type {
      display: inline;
    }
  }

  ul li {
    list-style-type: disc;
  }

  ul ul li {
    list-style-type: circle;
  }

  ul ul ul li {
    list-style-type: square;
  }

  ol li {
    list-style-type: decimal;
  }

  ol ol li {
    list-style-type: lower-latin;
  }

  ol ol ol li {
    list-style-type: lower-roman;
  }

  img {
    display: block;
    width: auto;
    max-height: 500px;
    cursor: pointer;
  }
}

.mention {
  font-weight: bold;
  color: var(--pneumatic-color-black72);
}

.video {
  /* stylelint-disable-next-line property-no-unknown */
  aspect-ratio: 16 / 9;
  width: 100%;
}

.loading-image {
  width: 100%;
  min-height: 64px;
  pointer-events: none;
  background-color: #f6f6f6;
  background-image: linear-gradient(90deg, #f6f6f6, white, #f6f6f6);
  background-repeat: no-repeat;
  background-size: 200px 100%;
  border-radius: 4px;
  animation: loading-animation 1.8s ease-in-out infinite;
}

@keyframes loading-animation {
  0% {
    background-position: -200px 0;
  }
  100% {
    background-position: calc(200px + 100%) 0;
  }
}

.checklist {
  margin: 16px 32px;

  @media (--mobile) {
    margin: 16px 0;
  }
}

.checklist_interactive {
  color: var(--pneumatic-color-black100);
}

.checkbox-wrapper {
  display: flex;
  min-height: 20px;

  &:not(:last-child) {
    margin-bottom: 8px;
  }
}

.checkbox {
  display: flex;
}

.checkbox-label {
  padding: 0;
}

.checkbox-content {
  margin-left: 28px;
}

.progressbar-container {
  margin-bottom: 6px;
}

.progressbar {
  display: flex;
  align-items: center;
}

.progressbar-hint {
  margin-left: 8px;
}

.progressbar-hint-icon {
  display: flex;
}

.checkbox-fake-placeholder {
  position: absolute;
  width: 20px;
  height: 20px;
  background-color: transparent;
  border: 1px solid var(--pneumatic-color-black16);
  border-radius: 4px;
}
