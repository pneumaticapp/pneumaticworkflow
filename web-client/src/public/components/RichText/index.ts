/* eslint-disable */
/* prettier-ignore */
export * from './RichText';
