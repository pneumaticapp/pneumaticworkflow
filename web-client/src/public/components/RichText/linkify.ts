/* eslint-disable */
/* prettier-ignore */
// @ts-nocheck
// tslint:disable

// This file was copied from https://github.com/jonschlinkert/remarkable/blob/58b6945f203ca7a0bb5a0785df90a3a6a8b9e59c/lib/linkify.js
// and then the link_open token params was changed to correctly work with editor attachments.

import Autolinker from 'autolinker';


let LINK_SCAN_RE = /www|@|\:\/\//;


function isLinkOpen(str) {
  return /^<a[>\s]/i.test(str);
}
function isLinkClose(str) {
  return /^<\/a\s*>/i.test(str);
}

// Stupid fabric to avoid singletons, for thread safety.
// Required for engines like Nashorn.
//
function createLinkifier() {
  let links = [];
  let autolinker = new Autolinker({
    stripPrefix: false,
    url: true,
    email: true,
    replaceFn (match) {
      // Only collect matched strings but don't change anything.
      switch (match.getType()) {
      /*eslint default-case:0*/
      case 'url':
        links.push({
          text: match.matchedText,
          url: match.getUrl(),
        });
        break;
      case 'email':
        links.push({
          text: match.matchedText,
          // normalize email protocol
          url: `mailto:${  match.getEmail().replace(/^mailto:/i, '')}`,
        });
        break;
      }

      return false;
    },
  });

  return {
    links,
    autolinker,
  };
}


function parseTokens(state) {
  let i; let j; let l; let tokens; let token; let text; let nodes; let ln; let pos; let level; let htmlLinkLevel;
  let blockTokens = state.tokens;
  let linkifier = null; let links; let autolinker;

  for (j = 0, l = blockTokens.length; j < l; j++) {
    if (blockTokens[j].type !== 'inline') {
      continue;
    }
    tokens = blockTokens[j].children;

    htmlLinkLevel = 0;

    // We scan from the end, to keep position when new tags added.
    // Use reversed logic in links start/end match
    for (i = tokens.length - 1; i >= 0; i--) {
      token = tokens[i];

      // Skip content of markdown links
      if (token.type === 'link_close') {
        i--;
        while (tokens[i].level !== token.level && tokens[i].type !== 'link_open') {
          i--;
        }
        continue;
      }

      // Skip content of html tag links
      if (token.type === 'htmltag') {
        if (isLinkOpen(token.content) && htmlLinkLevel > 0) {
          htmlLinkLevel--;
        }
        if (isLinkClose(token.content)) {
          htmlLinkLevel++;
        }
      }
      if (htmlLinkLevel > 0) {
        continue;
      }

      if (token.type === 'text' && LINK_SCAN_RE.test(token.content)) {

        // Init linkifier in lazy manner, only if required.
        if (!linkifier) {
          linkifier = createLinkifier();
          links = linkifier.links;
          autolinker = linkifier.autolinker;
        }

        text = token.content;
        links.length = 0;
        autolinker.link(text);

        if (!links.length) {
          continue;
        }

        // Now split string to nodes
        nodes = [];
        level = token.level;

        for (ln = 0; ln < links.length; ln++) {

          if (!state.inline.validateLink(links[ln].url)) {
            continue;
          }

          pos = text.indexOf(links[ln].text);

          if (pos) {
            nodes.push({
              type: 'text',
              content: text.slice(0, pos),
              level,
            });
          }
          nodes.push({
            type: 'link_open',
            url: links[ln].url,
            name: links[ln].text,
            isLinkified: true,
            level: level++,
          });
          nodes.push({
            type: 'link_close',
            level: --level,
          });
          text = text.slice(pos + links[ln].text.length);
        }
        if (text.length) {
          nodes.push({
            type: 'text',
            content: text,
            level,
          });
        }

        // replace current node
        blockTokens[j].children = tokens = [].concat(tokens.slice(0, i), nodes, tokens.slice(i + 1));
      }
    }
  }
};

export function linkify(md) {
  md.core.ruler.push('linkify', parseTokens);
};
