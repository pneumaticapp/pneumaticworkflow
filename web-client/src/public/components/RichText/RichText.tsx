/* eslint-disable */
/* prettier-ignore */
import * as React from 'react';
import ReactDOMServer from 'react-dom/server';
import * as xssFilters from 'xss-filters';
import { useDispatch } from 'react-redux';
import { Remarkable } from 'remarkable';
import { linkify } from './linkify';

import {
  mentionsRegex,
  youtubeVideoRegexp,
  loomVideoRegexp,
  wistiaVideoRegexp,
  variableRegex,
} from '../../constants/defaultValues';
import { isArrayWithItems } from '../../utils/helpers';
import { TTaskVariable } from '../TemplateEdit/types';
import { truncateString } from '../../utils/truncateString';
import { openFullscreenImage } from '../../redux/general/actions';
import { DocumentAttachment } from '../Attachments/DocumentAttachment';
import {
  linksRemarkablePlugin,
  checklistsRemarkablePlugin,
  TCustomLinkToken,
} from '../RichEditor/utils/сonverters/customMarkdownPlugins';
import { LinkCloseToken } from 'remarkable/lib';
import { ECustomEditorEntities } from '../RichEditor/utils/types';
import { createChecklistRenderer } from '../TaskCard/checklist/createChecklistRenderer';
import { createCheckPlaceholderId } from '../TaskCard/checklist/createCheckPlaceholderId';
import { prepareChecklistsForRendering } from '../../utils/checklists/prepareChecklistsForRendering';
import { createProgressbarPlaceholderId } from '../TaskCard/checklist/createProgressbarPlaceholderId';

import styles from './RichText.css';
import badgeStyles from '../../utils/badge/Badge.css';

export interface IRichTextProps {
  text: string | null;
  isMarkdownMode?: boolean;
  embedVideos?: boolean;
  variables?: TTaskVariable[];
  renderExtensions?: React.ReactNode[];
  interactiveChecklists?: boolean;
}

const MAX_VARIABLE_LENGTH = 20;

export function RichText({
  text,
  isMarkdownMode = true,
  embedVideos = true,
  variables,
  renderExtensions,
  interactiveChecklists,
}: IRichTextProps) {
  if (!text) {
    return null;
  }

  const [isRendered, setIsRendered] = React.useState(false);

  React.useLayoutEffect(() => {
    setIsRendered(true);
  }, []);

  const md = new Remarkable({
    html: true,
    breaks: true,
  })
    .use(checklistsRemarkablePlugin)
    .use(linkify)
    .use(attachmnetRenderer(embedVideos))
    .use(
      createChecklistRenderer({
        renderCheck: (listApiName, itemApiName) => {
          if (interactiveChecklists) {
            return `<div id="${createCheckPlaceholderId(listApiName, itemApiName)}"></div>`;
          }

          return `<div class="${styles['checkbox-fake-placeholder']}"></div>`;
        },
        renderProgressbar: (listApiName) => {
          if (!interactiveChecklists) {
            return '';
          }

          return `<div id="${createProgressbarPlaceholderId(listApiName)}"></div>`;
        },
        interactiveChecklist: interactiveChecklists,
      }),
    );

  const dispatch = useDispatch();
  const conatainerRef = React.useRef<HTMLDivElement>(null);
  const handleClick = React.useCallback((event: MouseEvent) => {
    const target = event.target as Element;
    const url = target.getAttribute('src');
    if (target.tagName === 'IMG' && url) {
      event.stopImmediatePropagation();
      dispatch(openFullscreenImage({ url }));
    }
  }, []);

  const showImagePlaceholders = (container: HTMLDivElement) => {
    const images = container.getElementsByTagName('img');
    for (let i = 0; i < images.length; i++) {
      images[i].classList.add(styles['loading-image']);
      images[i].onload = () => {
        images[i].classList.remove(styles['loading-image']);
      };
    }
  };

  React.useEffect(() => {
    if (conatainerRef.current) {
      showImagePlaceholders(conatainerRef.current);
      conatainerRef.current.addEventListener('click', handleClick);
    }

    return () => {
      if (conatainerRef.current) {
        conatainerRef.current.removeEventListener('click', handleClick);
      }
    };
  }, []);

  const REPLACE_RULES: {
    regExp: RegExp;
    // tslint:disable-next-line: no-any
    replaceLogic: any;
  }[] = [
    {
      // This is a hacky way to correctly render new lines in markdown
      regExp: /\n/g,
      replaceLogic: (match: string, offset: number, str: string) => {
        const nextSymbol = str[offset + 1];

        return nextSymbol === '\n' ? '\n&nbsp;\n' : match;
      },
    },
    {
      regExp: mentionsRegex,
      replaceLogic: `<span class='${styles['mention']}'>@$1</span>`,
    },
    {
      regExp: variableRegex,
      replaceLogic: (match: string, variableApiName: string) => {
        if (!isArrayWithItems(variables)) {
          return match;
        }

        const variable = variables.find((variable) => variableApiName === variable.apiName);
        if (!variable) {
          return match;
        }

        return `<span class="${badgeStyles['badge']} ${badgeStyles['specifity']}">${truncateString(
          variable.title,
          MAX_VARIABLE_LENGTH,
        )}</span>`;
      },
    },
  ];

  const sanitizedText = xssFilters.inHTMLData(prepareChecklistsForRendering(text));

  const htmlString = REPLACE_RULES.reduce((acc, { regExp, replaceLogic }) => {
    const replaceRegex = new RegExp(regExp, 'gi');

    return acc.replace(replaceRegex, replaceLogic);
  }, sanitizedText);

  if (!isMarkdownMode) {
    return <div dangerouslySetInnerHTML={{ __html: htmlString }} />;
  }

  return (
    <>
      {isRendered && renderExtensions}
      <div
        ref={conatainerRef}
        className={styles['container']}
        dangerouslySetInnerHTML={{ __html: md.render(htmlString) }}
      />
    </>
  );
}

const attachmnetRenderer = (embedVideoLinks: boolean) => {
  const renderLink = (url: string, name: string) => {
    const defaultLinkHtml = `<a href="${url}" target="_blank">${name}<a/>`;
    if (!embedVideoLinks) {
      return defaultLinkHtml;
    }

    const renderRules = [
      {
        regExp: loomVideoRegexp,
        replaceLogic: (match: string, group1: string) => {
          return `<div><iframe class="${styles['video']}" frameBorder="0" src="https://www.useloom.com/embed/${group1}" webkitallowfullscreen mozallowfullscreen allowfullscreen></iframe></div>`;
        },
      },
      {
        regExp: youtubeVideoRegexp,
        replaceLogic: (match: string, group1: string) => {
          return `<div><iframe class="${styles['video']}" frameBorder="0" src="//www.youtube.com/embed/${group1}" webkitallowfullscreen mozallowfullscreen allowfullscreen></iframe></div>`;
        },
      },
      {
        regExp: wistiaVideoRegexp,
        replaceLogic: (match: string, videoId: string) => {
          const videoUrl = `https://fast.wistia.com/embed/medias/${videoId}`;

          return `<div><iframe class="${styles['video']}" allowtransparency="true" frameborder="0" scrolling="no" src="${videoUrl}" webkitallowfullscreen mozallowfullscreen allowfullscreen></iframe></div>`;
        },
      },
    ];

    const renderVideoLogic = renderRules.find(({ regExp }) => regExp.test(url));
    if (renderVideoLogic) {
      return url.replace(renderVideoLogic.regExp, renderVideoLogic.replaceLogic);
    }

    return defaultLinkHtml;
  };

  return (md: Remarkable) => {
    md.inline.ruler.disable(['links']);
    md.use(linksRemarkablePlugin, {});

    // @ts-ignore
    md.renderer.rules.link_open = (tokens: TCustomLinkToken[], idx: number) => {
      const url = tokens[idx]['url'];
      const name = tokens[idx]['name'];
      const entityType = tokens[idx]['entityType'];
      const isLinkified = tokens[idx]['isLinkified'];

      if (isLinkified || !entityType) {
        return renderLink(url, name);
      }

      const renderMap: { [key in ECustomEditorEntities]: string } = {
        [ECustomEditorEntities.Link]: `<a href="${url}" target="_blank">`,
        [ECustomEditorEntities.Image]: `<img src=${url} />`,
        [ECustomEditorEntities.Video]: ReactDOMServer.renderToStaticMarkup(<video src={url} preload="auto" controls />),
        [ECustomEditorEntities.File]: ReactDOMServer.renderToStaticMarkup(
          <DocumentAttachment name={name} url={url} isEdit={false} />,
        ),
        [ECustomEditorEntities.Variable]: '',
        [ECustomEditorEntities.Mention]: '',
      };

      return renderMap[entityType];
    };

    // @ts-ignore
    md.renderer.rules.link_close = (tokens: LinkCloseToken[], idx: number) => {
      // @ts-ignore
      const entityType = tokens[idx]['entityType'];
      if (entityType === 'LINK') {
        return '</a>';
      }

      return '';
    };
  };
};
