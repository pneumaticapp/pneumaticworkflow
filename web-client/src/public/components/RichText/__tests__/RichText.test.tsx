/* eslint-disable */
/* prettier-ignore */
import * as React from 'react';
import { mount, render } from 'enzyme';
import { RichText, IRichTextProps } from '../RichText';

describe('RichText', () => {
  it('renderes correct html for mentions and links', () => {
    const props: IRichTextProps = {
      text: 'Hello, [Jyoti Puri|3], look here: link: http://pneumatic.app/',
      isMarkdownMode: true,
    };

    const wrapper = mount(<RichText {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renderes mentions in different languages correctly', () => {
    const users = [
      {
        name: 'Алексей Карпов-Константинов',
        id: 3,
      },
      {
        name: 'Alex Karpov-Konstantinov',
        id: 4,
      },
      {
        name: 'אלכסיי קרפוב-קונסטנטינוב',
        id: 5,
      },
      {
        name: 'アレクセイカルポフ-コンスタンティノフ',
        id: 6,
      },
    ];
    const usersString = users.map(({ name, id }) => `[${name}|${id}]`).join(', ');
    const props: IRichTextProps = { text: `Hello, ${usersString}`, isMarkdownMode: true, };

    const wrapper = mount(<RichText {...props} />);

    for (let user of users) {
      expect(wrapper.text()).toContain(`@${user.name}`);
    }
  });

  it('renderes correct html for plain text', () => {
    const props: IRichTextProps = { text: 'Some plain text', isMarkdownMode: true, };

    const wrapper = mount(<RichText {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('clears xss', () => {
    const props: IRichTextProps = {
      text: '<script>console.log("I\'ll try to hack you")</script>',
      isMarkdownMode: false,
    };

    const wrapper = mount(<RichText {...props} />);

    expect(wrapper.text()).toBe('<script>console.log("I\'ll try to hack you")</script>');
  });

  it('empty render if text prop is null', () => {
    const props: IRichTextProps = { text: null, isMarkdownMode: true, };

    const wrapper = mount(<RichText {...props} />);

    expect(wrapper.isEmptyRender()).toBe(true);
  });

  it('RichText and youtube in text flag return text with youtube iframe', () => {
    const props: IRichTextProps = {
      text: 'test1 https://www.youtube.com/watch?v=jNQXAC9IVRw',
      isMarkdownMode: true,
    };

    const wrapper = render(<RichText {...props}/>);

    expect(wrapper).toMatchSnapshot();
  });

  it('RichText and loom link in text flag return text with loom iframe', () => {
    const props: IRichTextProps = {
      text: 'test1 https://www.loom.com/share/7f68fa7f01e349cab91b0c36168f68c3?t=1',
      isMarkdownMode: true,
    };

    const wrapper = render(<RichText {...props}/>);

    expect(wrapper).toMatchSnapshot();
  });

  it('RichText and youtube and loom link in text flag return text with youtube and loom iframe',
    () => {
      const props: IRichTextProps = {
        text: 'test1 https://www.loom.com/share/7f68fa7f01e349cab91b0c36168f68c3?t=1 https://www.youtube.com/watch?v=jNQXAC9IVRw',
        isMarkdownMode: true,
      };

      const wrapper = render(<RichText {...props}/>);

      expect(wrapper).toMatchSnapshot();
    });
});
