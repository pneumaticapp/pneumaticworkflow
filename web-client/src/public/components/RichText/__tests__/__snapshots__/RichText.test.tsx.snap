// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RichText RichText and loom link in text flag return text with loom iframe 1`] = `
<div
  class="container"
>
  <p>
    test1 
  </p>
  <div>
    <iframe
      allowfullscreen=""
      class="video"
      frameborder="0"
      mozallowfullscreen=""
      src="https://www.useloom.com/embed/7f68fa7f01e349cab91b0c36168f68c3"
      webkitallowfullscreen=""
    />
  </div>
  <p />
  

</div>
`;

exports[`RichText RichText and youtube and loom link in text flag return text with youtube and loom iframe 1`] = `
<div
  class="container"
>
  <p>
    test1 
  </p>
  <div>
    <iframe
      allowfullscreen=""
      class="video"
      frameborder="0"
      mozallowfullscreen=""
      src="https://www.useloom.com/embed/7f68fa7f01e349cab91b0c36168f68c3"
      webkitallowfullscreen=""
    />
  </div>
   
  <div>
    <iframe
      allowfullscreen=""
      class="video"
      frameborder="0"
      mozallowfullscreen=""
      src="//www.youtube.com/embed/jNQXAC9IVRw"
      webkitallowfullscreen=""
    />
  </div>
  <p />
  

</div>
`;

exports[`RichText RichText and youtube in text flag return text with youtube iframe 1`] = `
<div
  class="container"
>
  <p>
    test1 
  </p>
  <div>
    <iframe
      allowfullscreen=""
      class="video"
      frameborder="0"
      mozallowfullscreen=""
      src="//www.youtube.com/embed/jNQXAC9IVRw"
      webkitallowfullscreen=""
    />
  </div>
  <p />
  

</div>
`;

exports[`RichText renderes correct html for mentions and links 1`] = `
<RichText
  isMarkdownMode={true}
  text="Hello, [Jyoti Puri|3], look here: link: http://pneumatic.app/"
>
  <div
    className="container"
    dangerouslySetInnerHTML={
      {
        "__html": "<p>Hello, <span class='mention'>@Jyoti Puri</span>, look here: link: <a href="http://pneumatic.app/" target="_blank">http://pneumatic.app/<a/></p>
",
      }
    }
  />
</RichText>
`;

exports[`RichText renderes correct html for plain text 1`] = `
<RichText
  isMarkdownMode={true}
  text="Some plain text"
>
  <div
    className="container"
    dangerouslySetInnerHTML={
      {
        "__html": "<p>Some plain text</p>
",
      }
    }
  />
</RichText>
`;
